OSDOCS 9447: Restructured oc-mirror section to include both V1 and V2 changes. (#74485)
openshift/openshift-docs: installing/disconnected_install/installing-mirroring-disconnected-v2.adoc

.Mirror to Disk

Pulls the container images from the source specified in the image set configuration and pack them into a tar archive on disk (local directory).

[source,terminal]
----
$ oc-mirror -c isc.yaml <file path> --v2
----

.Disk to Mirror

Copy the containers images from the tar archive to a container registry (--from flag is required on this workflow).

[source,terminal]
----
$ oc-mirror -c isc.yaml --from file:///oc-mirror/mirror1 docker://localhost:6000 --v2
----

.Mirror To Mirror

Copy the container images from the source specified in the image set configuration to the destination (container registry).

[source,terminal]
----
$ oc-mirror -c isc.yaml --workspace ffile:///oc-mirror/mirror1 docker://localhost:6000 --v2
----

[id="oc-mirror-delete"]
== Image delete workflow in oc-mirror

In V1, pruning (or deletion of images) is performed automatically at the end of the mirroring. In order to delete a specific index (release, catalog) or additionalImage with v1, the same "StorageConfig" setting should be used, and the imagesetconfig is updated omiting the content to be pruned. The pruning logic will find the "difference" and prune the images accordingly.

> This has its pros and cons, but due to the large amount of bugs that have been filed, we decided to change the behaviour in v2.

Author: Ack.

You can use the oc-mirror OpenShift CLI (`oc`) plugin to mirror images to a mirror registry in your fully or partially disconnected environments. You must run oc-mirror from a system with internet connectivity in order to download the required images from the official Red Hat registries.

[id="installation-oc-mirror-workflow_{context}"]
== High level workflow
The following steps outline the high-level workflow on how to use the oc-mirror plugin to mirror images to a mirror registry:

. Create an image set configuration file.
. Mirror the image set to the mirror registry by using one of the following methods:
** Mirror an image set directly to the mirror registry.
** Mirror an image set to disk, transfer the image set to the target environment, then upload the image set to the target mirror registry.
. Configure your cluster to use the resources generated by the oc-mirror plugin.
. Repeat these steps to update your mirror registry as necessary.

// About the oc-mirror plugin

> include::modules/oc-mirror-about.adoc[leveloffset=+1]

Author: Ack..

// About V2
include::modules/oc-mirror-v2-about.adoc[leveloffset=+2]

// oc-mirror compatibility and support
include::modules/oc-mirror-support.adoc[leveloffset=+1]

> // About the mirror registry

Reviewer: Suggestion: is it possible to put this in a annex section? 
In this position in the doc, it feels to me that we've already dug a bit into the oc-mirror specifics, to go back to discussing mirror registry. WDYT?

Reviewer: Maybe a link to it from the pre-requisites section below this?

Author: Yep, as discussed it will be moved.

====
+
If you do not already have an existing solution for a container image registry, subscribers of {product-title} are provided a link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.15/html/installing/disconnected-installation-mirroring#installing-mirroring-creating-registry[mirror registry for Red Hat OpenShift]. The _mirror registry for Red Hat OpenShift_ is included with your subscription and is a small-scale container registry that can be used to mirror the required container images of {product-title} in disconnected installations.

> * Save the file as `$XDG_RUNTIME_DIR/containers/auth.json`.

Reviewer: I think this is misplaced, maybe from environment preparation?

Author: line 53?

Reviewer 2: Yes its mentioned in line 69 below - so maybe delete this ?

[id="using-oc-mirror"]
== Using oc-mirror

The instructions in this section will show how to prepare the environment and use oc-mirror.

.Environment Preparation

* To set up the credentials for pulling/pushing container images, download the pull secret and place it in the correct path.

** The default podman credentials location `$XDG_RUNTIME_DIR/containers/auth` is used for authenticating to the registries.
** The docker location `~/.docker/config.json` for credentials is also supported.

[Optional Step] 
* Use the following command to run a local registry that will be used as a destination:

[source,terminal]
----

> $ podman run -dt -p 6000:5000 -e REGISTRY_STORAGE_DELETE_ENABLED=true --name local-registry docker.io/library/registry:2

Reviewer: not sure we need this optional step, especially that there is a lot of content in prerequisite, and aboute the mirror registry to explain how to setup a highly available registry.
we don' t want to tell users to have a local (not hightly available) registry for their production mirroring.

Reviewer 3: +1, i agree

Author: Ack

Below is an example which included an Openshift Release (version 4.13.10), an operator catalog with only 3 operators and some additional images (ubi8 and ubi9):

[source,yaml]
----
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v2alpha1
mirror:
  platform:
    channels:
    - name: stable-4.13
      minVersion: 4.13.10
      maxVersion: 4.13.10
    graph: true
  operators:
    - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.15
      packages:
       - name: aws-load-balancer-operator
       - name: 3scale-operator
       - name: node-observability-operator
  additionalImages: 
   - name: registry.redhat.io/ubi8/ubi:latest
   - name: registry.redhat.io/ubi9/ubi@sha256:20f695d2a91352d4eaa25107535126727b5945bff38ed36a3e59590f495046f0
----

More Image Set Configuration example can be found in the additional resources.

[id="oc-mirror-workflows"]
== oc-mirror workflows

This section will explain all the workflows supported by oc-mirror currently.

.Mirror to Disk

Pulls the container images from the source specified in the image set configuration and pack them into a tar archive on disk (local directory).

[source,terminal]
----

> $ oc-mirror -c isc.yaml <file path> --v2

Reviewer 3: replace <file path> with file:///oc-mirror/mirror1 ?

Author: Done

----

.Disk to Mirror

> Copy the containers images from the tar archive to a container registry (--from flag is required on this workflow).

Reviewer 3: @subhtk I am not sure if the wording here is right, i think these needs to be changed/modified a bit as per the documentation standarads.

Author: I will rephrase it.

[source,terminal]
----
$ oc-mirror -c isc.yaml --from file:///oc-mirror/mirror1 docker://localhost:6000 --v2
----

.Mirror To Mirror

Copy the container images from the source specified in the image set configuration to the destination (container registry).

[source,terminal]
----

> $ oc-mirror -c isc.yaml --workspace ffile:///oc-mirror/mirror1 docker://localhost:6000 --v2

Reviewer 3: Typo in the command file. s/ffile/file.

@sherine-k do we need to add --dest-tls-verify=false as well or do you think user or customer would know based on the error they receive ?

Reviewer 2: We should always encourage https over http , so maybe wording to say use --help if using an insecure registry

Author: Done

----

[id="oc-mirror-delete"]
== Image delete workflow in oc-mirror

> In V1, pruning (or deletion of images) is set automatically, if a specific index (release, catalog) or additionalImage is to be deleted, the current workflow would be to use the same "StorageConfig" setting and omit the previously mirrored images, the pruning logic will find the "difference" and prune the images accordingly.

Reviewer: ```suggestion
In V1, pruning (or deletion of images) is performed automatically at the end of the mirroring. In order to delete a specific index (release, catalog) or additionalImage with v1, the same "StorageConfig" setting should be used, and the imagesetconfig is updated omiting the content to be pruned. The pruning logic will find the "difference" and prune the images accordingly.
```

Author: Ack.

> This has its pros and cons, but due to the large amount of bugs been filed, we decided to change the behaviour in v2.

Reviewer: ```suggestion
This has its pros and cons, but due to the large amount of bugs that have been filed, we decided to change the behaviour in v2.
```

Author: Ack.

> In v2 a new kind called "DeleteImageSetConfiguration" has been introduced, this is used to configure the deletion of images, to avoid using an ImageSetConfig, accidentally deleting wanted (deployed) images.

Reviewer: ```suggestion
In v2 a new kind called "DeleteImageSetConfiguration" has been introduced, this is used to configure the deletion of images, to avoid using an ImageSetConfig, accidentally deleting useful (deployed) images.
```

Reviewer 3: s/wanted/required

Author: Ack.

A new oc-mirror delete sub command is now available to execute the delete workflow with the DeleteImageSetConfiguration input.

> This functionality is able to work in a complete "air-gapped" (disconnected) environment, provided that the images have been previously mirrored (i.e the local cache and remote registry know about the images being deleted).

Reviewer 3: s/ is able to work / should work

Author: Let me rephrase this.

The delete functionality is split into 2 stages:

- The first stage is to create (generate) a delete yaml file using the --generate flag, this file is used to validate the images/blobs that will be deleted.
- The second stage is to use the delete yaml file (once validated) to actually perform a local cache and remote registry delete.

.For stage 1:
** Use the following command to create `delete` yaml file for delete phase:

[source,terminal]
----
$ oc-mirror delete --config delete-image-set-config.yaml --workspace file://<previously-mirrored-work-folder> --v2 --generate docker://<remote-registry> 
----

** Use the following command to compare versions:

[source,terminal]
----
$ oc-mirror delete --config delete-image-set-config.yaml --workspace file://<previously-mirrored-work-folder> --v2 --generate --delete-id v4.11 docker://<remote-registry>

> ----

Reviewer 3: It is not very clear on why some one needs to to this. It would be great if you could explain on why we are doing this

Author: Yes, as discussed I will be adding the necessary information.

.For stage 2:
** Once the generate stage is completed, use the following command to delete the images from the remote registry. 

[source,terminal]
----
$ oc-mirror delete --v2 --delete-yaml-file <previously-mirrored-work-folder>/delete/delete-images-v4.11.yaml docker://<remote-registry>  
----

** Use the following command to delete the remote registry and forces the cache delete:

[source,terminal]
----
$ oc-mirror delete --v2 --delete-yaml-file <previously-mirrored-work-folder>/delete/delete-images-v4.11.yaml --force-cache-delete true docker://<remote-registry> 

> ----

Reviewer 3: Not very sure about these steps, needs to see if the above stages works for all kind of mirroring i.e catalog, oci based and mirror2mirror

Author: Okay.

Here is an example of the `DeleteImageSetConfig`:

[source,yaml]
----

> apiVersion: mirror.openshift.io/v1alpha2

Reviewer: ```suggestion
apiVersion: mirror.openshift.io/v2alpha1
```

Reviewer 3: Example needs to be at the beginning where we are introducing DeleteImageSetConfig.

Author: Ack.

kind: DeleteImageSetConfiguration
delete:
  platform:
    channels:
      - name: stable-4.13 
        minVersion: 4.13.3
        maxVersion: 4.13.3
  operators:
    - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.12
      packages:

>       - name: aws-load-balancer-operator

Reviewer: ```suggestion
      - name: aws-load-balancer-operator
         minVersion: 0.0.1
         maxVersion: 0.0.1
```

  additionalImages: 
    - name: registry.redhat.io/ubi8/ubi@sha256:bce7e9f69fb7d4533447232478fd825811c760288f87a35699f9c8f030f2c1a6
    - name: registry.redhat.io/ubi8/ubi-minimal@sha256:8bedbe742f140108897fb3532068e8316900d9814f399d676ac78b46e740e34e

> ----

Reviewer: With this deleteImageSetConfig, oc-mirror will remove:
* all images of OCP release 4.13.3, 
* the catalog image redhat-operator-index v4.12
* the latest aws-load-balancer-operator bundle and all its related images
* as well as the additional images ubi and ubi-minimal referenced by their corresponding digests.

BTW, for this example, `aws-load-balancer-operator` it is best to specify a min and max version here, otherwise, it might not even find the operator to be deleted...

Author: Understood.

** Save the file as `$XDG_RUNTIME_DIR/containers/auth.json`.

[id="building-image-set-configuration"]
== Building the image set configuration

The Image Set Configuration is the input file used by oc-mirror to discover which images are necessary to be mirrored.

Below is an example for the `ImageSetConfiguration` input file:

[source,yaml]
----
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v2alpha1
mirror:
  platform:
    channels:
    - name: stable-4.13
      minVersion: 4.13.10
      maxVersion: 4.13.10
    graph: true
  operators:
    - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.15
      packages:
       - name: aws-load-balancer-operator
       - name: 3scale-operator
       - name: node-observability-operator
  additionalImages: 
   - name: registry.redhat.io/ubi8/ubi:latest
   - name: registry.redhat.io/ubi9/ubi@sha256:20f695d2a91352d4eaa25107535126727b5945bff38ed36a3e59590f495046f0
----

More Image Set Configuration examples can be found in the additional resources.

[id="oc-mirror-workflows"]
== oc-mirror workflows

This section will explain all the workflows supported by oc-mirror currently.

> .Mirror to Disk

Reviewer: ```suggestion
.Mirror to Disk + Disk to Mirror

You can use this workflow when the target disconnected registry is fully disconnected. In other words, the environment on which you are executing oc-mirror has access to the internet but doesn't have access to the target disconnected registry.
In this case, the workflow is composed of 3 steps:
* Mirror to disk: preparing an archive containing the image set that needs to be mirrored
* A manual step, where the user transfers the archive to the network of the disconnected registry
* Disk to mirror: from an environment that has access to the disconnected registry, execute oc-mirror to mirror the image set from the archive to the target disconnected registry.
```

Author: Changed as discussed.

Pulls the container images from the source specified in the image set configuration and pack them into a tar archive on disk (local directory).

[source,terminal]
----
$ oc-mirror -c isc.yaml <file path> --v2
----

.Disk to Mirror

Copies the container images from the tar archive to a container registry.

[source,terminal]
----
$ oc-mirror -c isc.yaml --from file:///oc-mirror/mirror1 docker://localhost:6000 --v2
----

.Mirror To Mirror

Copies the container images from the source specified in the image set configuration to the destination (container registry).

[source,terminal]
----
$ oc-mirror -c isc.yaml --workspace file:///oc-mirror/mirror1 docker://localhost:6000 --v2
----

[id="oc-mirror-delete"]

> == Image delete workflow in oc-mirror

Reviewer 4: It is important to add an `Important Note` here that in v2 auto pruning is not happening anymore. Auto pruning means that in v1 when a customer removed something (an operator for example) from the image set configuration, this operator was going to be pruned from their container registry automatically. 

A lot of customers complained about this auto pruning and that's why in v2 we have a specific file named DeleteImageSetConfiguration to include only the images that the customer wants to delete from their registry. And the customer needs to run the delete command using the Delete file. In this way we avoid customers deleting things from their container registries by mistake.

Author: Added

> == Image delete workflow in oc-mirror

Reviewer 4: It is also important to mention in the Image delete workflow that oc-mirror only delete the manifests of the images (this does not reduce the storage being occupied in the registry). 

In order to reduce the storage of unnecessary images (the ones with manifests deleted) the customer needs to enable the garbage collector on their container registry (the procedure on doing that depends on which container registry they are using). Only with the garbage collector enabled the container registry will delete the images (blobs) which does not have reference to any manifest, reducing the storage previous occupied by the blobs deleted.

See the comment above as requested by you @kasturinarra.

The Image Set Configuration is the input file used by oc-mirror to discover which images are necessary to be mirrored.

Below is an example for the `ImageSetConfiguration` input file:

[source,yaml]
----
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v2alpha1
mirror:
  platform:
    channels:
    - name: stable-4.13
      minVersion: 4.13.10
      maxVersion: 4.13.10
    graph: true
  operators:
    - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.15
      packages:
       - name: aws-load-balancer-operator
       - name: 3scale-operator
       - name: node-observability-operator
  additionalImages: 
   - name: registry.redhat.io/ubi8/ubi:latest
   - name: registry.redhat.io/ubi9/ubi@sha256:20f695d2a91352d4eaa25107535126727b5945bff38ed36a3e59590f495046f0
----

[id="oc-mirror-workflows-partially-disconnected"]
=== Mirroring an image set in a partially disconnected environment

.Mirror To Mirror

You can use this workflow for partially disconnected mirroring.
You can use this workflow when the environment from which you are executing oc-mirror has both:

- Access to the internet
- Access to the target disconnected registry

> You are required to specify a storage backend in the image set configuration file. This storage backend can be a local directory or a Docker v2 registry. The oc-mirror plugin stores metadata in this storage backend during image set creation.

Reviewer 4: storage backend is not present in the oc-mirror v2

Author: Then for V2 what will be the change here?

Author: deleted

.Prerequisites

* You have access to the internet to get the necessary container images.
* You have installed the OpenShift CLI (`oc`).
* You have installed the `oc-mirror` CLI plugin.
* You have created the image set configuration file.

.Procedure

Copy the container images from the source specified in the image set configuration to the destination (container registry).

[source,terminal]
----
$ oc mirror -c isc.yaml --workspace file://oc-mirror/mirror1 docker://localhost:6000 --v2
----

.Verification

> . Navigate into the `oc-mirror-workspace/` directory that was generated.

Reviewer 4: Instead of `oc-mirror-workspace` it is called `working-dir` under the path that you specified on --workspace

> . Navigate into the results directory, for example, `results-1639608409/`.

Reviewer 4: here instead of `results-1639608409/` you can change to `cluster-resources`

* A manual step, where the user transfers the archive to the network of the disconnected registry
* Disk to mirror: from an environment that has access to the disconnected registry, execute oc-mirror to mirror the image set from the archive to the target disconnected registry.

.Mirroring from mirror to disk

You can use the oc-mirror plugin to generate an image set and save the contents to disk. The generated image set can then be transferred to the disconnected environment and mirrored to the target registry.

Pulls the container images from the source specified in the image set configuration and pack them into a tar archive on disk (local directory).

.Prerequisites

* You have access to the internet to get the necessary container images.
* You have installed the OpenShift CLI (`oc`).
* You have installed the `oc-mirror` CLI plugin.
* You have created the image set configuration file.

.Procedure

[source,terminal]
----
$ oc mirror -c isc.yaml 
<file path> <1>
--v2
----
<1> Add the desired file path 

.Mirroring from disk to mirror

Copies the container images from the tar archive to a container registry.

[source,terminal]
----

> $ oc mirror -c isc.yaml --from file://oc-mirror/mirror1 docker://localhost:6000 --v2

Author: Will those be same as v1?

> $ oc mirror -c isc.yaml --from file://oc-mirror/mirror1 docker://localhost:6000 --v2

Reviewer 4: @sherine-k suggested the content in our google docs for v2. You can get from there.

----

[id="oc-mirror-delete"]
== Deleting the images from disconnected environment

> In v1, image pruning or deletion occurs automatically at the end of the mirroring process. To delete a specific release, catalog, or `additionalImage`, use the same `StorageConfig` settings. Update the `imagesetconfig` to exclude the content you wish to prune. The pruning logic will identify the differences and remove the images accordingly. Since this method has both advantages and disadvantages, it was modified in v2.

Reviewer 4: Keeping v1 information in v2 docs is not making it confuse to the customers to understand?

In v2, a new input type called `DeleteImageSetConfiguration` has been introduced. This is used to configure image deletion, preventing the accidental deletion of required or deployed images by avoiding the use of an `ImageSetConfig`.

The new sub command allows the users to execute the delete workflow with the `DeleteImageSetConfiguration`.

This functionality works in a fully disconnected environment, provided the images have been previously mirrored. This means that both the local cache and the remote registry are aware of the images being deleted.

[NOTE]
====
In v2, automatic pruning is no longer performed. In v1, when an item (such as an operator) was removed from the image set configuration, it would be automatically pruned from the container registry. To prevent unintended deletions, v2 introduces a specific file named DeleteImageSetConfiguration. 

This file allows specifying only the images to be deleted from the registry. Users must run the delete command using this file, avoiding accidental deletions from their container registries.
====

> Using the DeleteImageSetConfiguration, oc-mirror will remove:
> - All images of {product-title} release 4.13.3
> - The catalog image `redhat-operator-index` v4.12
> - The latest `aws-load-balancer-operator` bundle and all its related images
> - The additional images `ubi` and `ubi-minimal` referenced by their corresponding digests.
> 
> Here is an example of the `DeleteImageSetConfig`:

Reviewer 4: ```suggestion
Using the DeleteImageSetConfiguration below as an example,  oc-mirror will remove: 

- All images of OpenShift Container Platform release 4.13.3 

- The catalog image redhat-operator-index v4.12 

- The aws-load-balancer-operator v0.0.1 bundle and all its related images 

- The additional images ubi and ubi-minimal referenced by their corresponding digests.
```

[id="prerequisites_oc-mirror-v2"]
== Prerequisites

* You must have a container image registry that supports link:https://docs.docker.com/registry/spec/manifest-v2-2[Docker v2-2] in the location that will host the {product-title} cluster, such as Red Hat Quay.
+
[NOTE]
====
If you use Red Hat Quay, you must use version 3.6 or later with the oc-mirror plugin. If you have an entitlement to Red Hat Quay, see the documentation on deploying Red Hat Quay link:https://access.redhat.com/documentation/en-us/red_hat_quay/3/html/deploy_red_hat_quay_for_proof-of-concept_non-production_purposes/[for proof-of-concept purposes] or link:https://access.redhat.com/documentation/en-us/red_hat_quay/3/html/deploying_the_red_hat_quay_operator_on_openshift_container_platform/index[by using the Red Hat Quay Operator]. If you need additional assistance selecting and installing a registry, contact your sales representative or Red Hat Support.
====

* Every machine in the provisioned clusters must have access to the mirror registry. If the registry is unreachable, tasks like installation, updating, or routine operations such as workload relocation may fail. 

> Therefore, mirror registries must be operated in a highly available manner, ensuring their availability aligns with the production availability of your {product-title} clusters.

Reviewer 2: @subhtk -  should this not be part of the same paragraph above ? (seems like it got misaligned)

* You have set the umask parameter to 0022 on the operating system that uses oc-mirror.

[id="using-oc-mirror"]
== Mirroring an image set to a mirror registry

> Oc-mirror workflows are for when the target disconnected registry is fully or partially disconnected from the internet. In other words, the environment on which you are executing oc-mirror has access to the internet but doesn't have (or have partially) access to the target disconnected registry.

Reviewer 2: @subhtk - in our v2 google doc this section was deletd

=== Building the image set configuration

The Image Set Configuration is the input file used by oc-mirror to discover which images are necessary to be mirrored.

Below is an example for the `ImageSetConfiguration` input file:

[source,yaml]
----
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v2alpha1
mirror:
  platform:
    channels:
    - name: stable-4.13
      minVersion: 4.13.10
      maxVersion: 4.13.10
    graph: true
  operators:
    - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.15
      packages:
       - name: aws-load-balancer-operator
       - name: 3scale-operator
       - name: node-observability-operator
  additionalImages: 
   - name: registry.redhat.io/ubi8/ubi:latest
   - name: registry.redhat.io/ubi9/ubi@sha256:20f695d2a91352d4eaa25107535126727b5945bff38ed36a3e59590f495046f0
----

[id="oc-mirror-workflows-partially-disconnected"]
=== Mirroring an image set in a partially disconnected environment

.Mirror To Mirror

You can use this workflow for partially disconnected mirroring, that is, when the environment from which you are executing oc-mirror has both:
- Access to the internet
- Access to the mirror registry

.Prerequisites

> * You have access to the internet to get the necessary container images.

Reviewer 2: @subhtk - also this section was deleted in the v2 doc

* Configure your cluster to use the resources generated by oc-mirror.

[id="oc-mirror-workflows-fully-disconnected"]
=== Mirroring an image set in a fully disconnected environment: Mirror-to-Disk and Disk-to-Mirror

In a fully disconnected environment, the mirror registry that the Openshift cluster can access is not reachable from the public internet. Following are the steps of the workflow:

* Mirror to disk: preparing an archive containing the image set that needs to be mirrored. This step requires internet access.
* A manual step, where the user transfers the archive to the network of the mirror registry (disconnected registry)
* Disk to mirror: from the environment that has access to the mirror registry, execute oc-mirror to mirror the image set from the archive to the target disconnected registry.

.Mirroring from mirror to disk

You can use the oc-mirror plugin to generate an image set and save the contents to disk. The generated image set can then be transferred to the disconnected environment and mirrored to the target registry.

Pulls the container images from the source specified in the image set configuration and pack them into a tar archive on disk (local directory).

> .Prerequisites

Reviewer 2: @subhtk - same here - also delete in v2 docs

<1> Add the desired file path 

.Mirroring from disk to mirror

Copies the container images from the tar archive to a container registry.

[source,terminal]
----
$ oc mirror -c isc.yaml --from file://oc-mirror/mirror1 docker://localhost:6000 --v2
----

[id="oc-mirror-delete"]
== Deleting the images from disconnected environment

In v2, a new input type called `DeleteImageSetConfiguration` has been introduced. This is used to configure image deletion, preventing the accidental deletion of required or deployed images by avoiding the use of an `ImageSetConfig`.

The new sub command allows the users to execute the delete workflow with the `DeleteImageSetConfiguration`.

This functionality works in a fully disconnected environment, provided the images have been previously mirrored. This means that both the local cache and the remote registry are aware of the images being deleted.

[NOTE]
====
In v2, automatic pruning is no longer performed. In v1, when an item (such as an operator) was removed from the image set configuration, it would be automatically pruned from the container registry. To prevent unintended deletions, v2 introduces a specific file named DeleteImageSetConfiguration. 

> This file allows specifying only the images to be deleted from the registry. Users must run the delete command using this file, avoiding accidental deletions from their container registries.

Reviewer 2: @subhtk - and same here (deleted in v2 docs)
